packages/lesswrong/lib/collections/users/helpers.tsx

```diff
@@ -300,8 +300,12 @@ export const userUseMarkdownPostEditor = (user: UsersCurrent|null): boolean => {
   return user.markDownPostEditor
 }
 
-export const userCanEdit = (currentUser, user) => {
-  return userOwns(currentUser, user) || userCanDo(currentUser, 'users.edit.all')
+export const userCanEditUser = (currentUser: UsersCurrent|DbUser|null, user: HasIdType|HasSlugType|UsersMinimumInfo|DbUser) => {
+  // We allow users to call this function with basically "pretend" user objects
+  // as the second argument. We know from inspecting userOwns that those pretend
+  // user objects are safe, but if userOwns allowed them it would make the type
+  // checks much less safe.
+  return userOwns(currentUser, user as UsersMinimumInfo|DbUser) || userCanDo(currentUser, 'users.edit.all')
```

Comment:
I think the real problem here is that the type of `userOwns` is just wrong. Looking at the implementation, I think it should just be something like:
```
export type OwnableDocument = Partial<Pick<HasUserIdType, "userId"> | Pick<DbUser, "slug" | "_id">>;

export const userOwns = function (user: UsersMinimumInfo|DbUser|null, document: OwnableDocument|null): boolean {
```
then this function can be
```
export const userCanEditUser = (currentUser: UsersCurrent|DbUser|null, user: OwnableDocument) => {
```
and we can remove the casts here and in `UserEditForm` and `EditProfileForm`.

Reply:
I thought about that. But the type of OwnableDocument is really, [morally](https://qr.ae/pvQ7Nk), `HasUserId | AUserObjectWithSlugOrId`. The reason I did not do basically what you describe is that AUserObjectWithSlugOrId, must actually be a *User* object. It should be a type error to pass in an object, like a tag, which has a slug, and an _id, but is not a user object, and does not have a userId.

Reply:
I agree it should morally be a `User`, but if that's a problem we really care about I don't think this solves it. Without actually checking because I'm lazy, I'm pretty sure you can already pass a Tag into this function because of `HasIdType`, no?

Reply:
You're right that UserOwns currently accepts DbObject. It shouldn't though. I've now stopped it from doing so, with no type errors.

Reply:
Sorry - that's not what I meant. I meant that you can do
```
const aTag = await Tags.find({}, {limit: 1});
return userCanEditUser(currentUser, aTag);
```
without type errors, which violates the problem you were trying to fix above.

Reply:
userCanEditUser has to accept HasIdType, due to the artificially constructed user in `EditProfileForm.tsx`. (I could make the type-casting happen there, perhaps.) This seems relatively ok to me because userCanEditUser now has a very clear name.

userOwns is trickier. There I'm more concerned about accidental misuse. Because it's less obvious what the type of the second argument is allowed to be, I think it's more important to get it right.